openvm/cuda/src/apc_tracegen.cu

```diff
@@ -0,0 +1,162 @@
+#include "primitives/buffer_view.cuh"
+#include "primitives/constants.h"
+#include "primitives/trace_access.h"
+#include "expr_eval.cuh"
+
+// ============================================================================================
+// Types
+// ============================================================================================
+
+struct OriginalAir {
+    int width;               // number of columns
+    int height;              // number of rows (Ha)
+    const Fp* buffer;        // column-major base: col*height + row
+    int row_block_size;      // stride between used rows
+    int substitutions_offset;// offset in d_subs
+    int substitutions_length;// count in d_subs for this AIR
+};
+
+struct Subst {
+    int col;      // source column within this AIR
+    int row;      // base row offset within the row-block
+    int apc_col;  // destination APC column
+};
+
+struct DerivedColumn {
+    int index; // destination APC column index
+    Fp  value; // constant value for valid rows
+};
+
+extern "C" {
+  typedef struct {
+    uint64_t col_base; // precomputed destination base offset = apc_col_index * H
+    DevArgSpan span;   // expression span encoding this column's value
+  } DerivedExprSpec;
+}
+
+// ============================================================================================
+// Kernel: one block per OriginalAir; each warp handles one substitution (APC column).
+// ============================================================================================
+
+__global__ void apc_tracegen_kernel(
+    Fp* __restrict__ d_output,                         // [output_height * output_width], column-major
+    const OriginalAir* __restrict__ d_original_airs,   // metadata per AIR
+    const Subst* __restrict__ d_subs,                  // all substitutions
+    size_t output_height,                              // H_out
+    int num_apc_calls                                  // number of APC calls
+) {
+    const int air_id = blockIdx.x;
+    const OriginalAir air = d_original_airs[air_id];
+
+    const Fp* __restrict__ src_base = air.buffer;
+    const int Ha  = air.height;
+    const int RBS = air.row_block_size;
+
+    const int lane  = threadIdx.x & 31;     // 0..31
+    const int warp  = threadIdx.x >> 5;     // warp index in block
+    const int warps_per_block = blockDim.x >> 5;
+
+    // Process this AIR's substitutions in batches of warps_per_block
+    for (int base = 0; base < air.substitutions_length; base += warps_per_block) {
```

Comment:
I think the `for` update clause should be `base += gridDim.x * warps_per_block`.
Also, same as in the other cuda file, the `warp` sum can be part of the `for` init clause i think